File: src/main/java/fr/skyost/serialkey/listener/LostChestsListener.java
package fr.skyost.serialkey.listener;

import org.bukkit.Location;
import org.bukkit.block.Chest;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.inventory.InventoryHolder;

import fr.skyost.serialkey.SerialKey;

/**
 * A listener that allows to listen chests related events.
 */

public class LostChestsListener extends SerialKeyListener {

	/**
	 * Creates a new lost chests listener instance.
	 *
	 * @param plugin The plugin.
	 */

	public LostChestsListener(final SerialKey plugin) {
		super(plugin);
	}

	@EventHandler(priority = EventPriority.HIGHEST)
	private void onInventoryClick(final InventoryClickEvent event) {
		if(!api.isUsedKey(event.getCurrentItem())) {
			return;
		}

		final InventoryHolder holder = event.getInventory().getHolder();
		if(!(holder instanceof Chest)) {
			return;
		}

		final Location location = ((Chest)holder).getLocation();
		if(api.hasPadlock(location) && api.extractLocation(event.getCurrentItem()).equals(location)) {
			event.setCancelled(true);
			plugin.sendMessage(event.getWhoClicked(), plugin.getPluginMessages().message6);
		}
	}

}

File: core/src/main/java/fr/skyost/serialkey/core/listener/LostChestsListener.kt
package fr.skyost.serialkey.core.listener;

import fr.skyost.serialkey.core.SerialKeyPlugin;
import fr.skyost.serialkey.core.object.SerialKeyLocation;
import fr.skyost.serialkey.core.object.SerialKeyPerson;

/**
 * A listener that allows to listen chests related events.
 */

public class LostChestsListener<I, L> extends SerialKeyListener<I, L> {

	/**
	 * Creates a new lost chests listener instance.
	 *
	 * @param plugin The plugin.
	 */

	public LostChestsListener(final SerialKeyPlugin<I, L> plugin) {
		super(plugin);
	}

	/**
	 * Triggered when a click occurs in an inventory.
	 *
	 * @param player The involved player.
	 * @param chestLocation The chest location.
	 * @param item The clicked item.
	 * @param cancelEvent The runnable that cancels the event.
	 */

	protected void onInventoryClick(final SerialKeyPerson player, final SerialKeyLocation chestLocation, final I item, final Runnable cancelEvent) {
		if(!itemManager.isUsedKey(item) && !itemManager.isUsedBunchOfKeys(item)) {
			return;
		}


		padlockManager.fixLocation(chestLocation);
		if(!padlockManager.hasPadlock(chestLocation, false)) {
			return;
		}

		for(final SerialKeyLocation location : unlocker.getLocations(item)) {
			padlockManager.fixLocation(location);
			if(location.equals(chestLocation)) {
				cancelEvent.run();
				plugin.sendMessage(player, plugin.getPluginMessages().getChestProtectionMessage());
				return;
			}
		}
	}

}